pkg/queue/bounded_queue.go

```diff
@@ -97,17 +105,52 @@ func (q *BoundedQueue) Stop() {
 	atomic.StoreInt32(&q.stopped, 1) // disable producer
 	close(q.stopCh)
 	q.stopWG.Wait()
+
+	// if we are resizing the queue, wait for it to complete
+	q.resizeLock.RLock()
+	defer q.resizeLock.RUnlock()
+
 	close(q.items)
 }
 
+// Resize the queue to the new capacity, returning whether a resize operation took place.
+// Use with care, as it will block all other important operations until the resize operation completes.
+func (q *BoundedQueue) Resize(newCap int) bool {
+	// if the capacity size hasn't changed, there's no need to resize
+	if cap(q.items) == newCap {
+		return false
+	}
+
+	// we need exclusive rw access to the internal queue
+	q.resizeLock.Lock()
+	defer q.resizeLock.Unlock()
+
+	new := make(chan interface{}, newCap)
```

Comment:
* what if newCap is less than currentCap and current size? This will probably deadlock
* please don't use `new` keyword as var name

Reply:
Funny, I thought I had added a test for the downscale, but looks like I haven't. Will add one.